lib/matplotlib/figure.py

```diff
@@ -1576,6 +1576,110 @@ def subplots(self, nrows=1, ncols=1, sharex=False, sharey=False,
             # Returned axis array will be always 2-d, even if nrows=ncols=1.
             return axarr
 
+    def build_grid(self, layout, *, subplot_kw=None, gridspec_kw=None,
+                   empty_sentinel=None):
+        """Build a layout based on an grid of strings.
+
+        This is a helper function to build complex GridSpec layouts visually.
+
+        Parameters
+        ----------
+        layout : list of list of {str, int,  Axes} or nested
+
+            A visual layout of how you want your Axes to be arranged
+            labeled as strings.  For example ::
+
+               x = [['A', 'A', 'B'],
+                    ['C', 'D', 'B']]
+
+            Produces 4 axes:
+
+            - 'A' which is 1 row high and spans the first two columns
+            - 'B' which is 2 rows high and is on the right edge
+            - 'C' which in 1 row and 1 column wide in the bottom left
+            - 'D' which in 1 row and 1 column wide in the bottom center
+
+        subplot_kw : dict, optional
+            Dict with keywords passed to the
+            :meth:`~matplotlib.figure.Figure.add_subplot` call used to create
+            each subplot.
+
+        gridspec_kw : dict, optional
+            Dict with keywords passed to the
+            `~matplotlib.gridspec.GridSpec` constructor used to create
+            the grid the subplots are placed on.
+
+        empty_sentinel : object, optional
+            Entry in the layout to mean "leave this space empty"
+
+
+        Returns
+        -------
+        dict[str, Axes]
+           A dictionary mapping the string labels to the new Axes objects.
+
+        """
+        subplot_kw = subplot_kw or {}
+        gridspec_kw = gridspec_kw or {}
+
+        def _process_layout(layout):
+            unique_ids = set()
+            nested = {}
+            for j, row in enumerate(layout):
+                for k, v in enumerate(row):
+                    if v == empty_sentinel:
+                        continue
+                    elif isinstance(v, (int, str, Axes)):
```

Comment:
if you're also going to support ints, we may as well support any (hashable) key?  for example, right now this doesn't support np.uint8 or np.str_ which I'm sure will come up.  I don't think just keying off whatever the user puts in is too onerous...

Reply:
I very briefly tried that but it got messier than I expected very quickly.  I don't clearly remember the way in which it got messy so I'll try again.

Reply:
got this working.